std/exception.d

```diff
@@ -2148,3 +2148,65 @@ pure nothrow @safe unittest
 
     auto infSlice = infinite[0 .. $]; // this would throw otherwise
 }
+
+
+/++
+    Convenience mixin for trivially sub-classing exceptions
+
+    Even trivially sub-classing an exception involves writing boilerplate code
+    for the constructor to: 1) correctly pass in the source file and line number
+    the exception was thrown from; 2) be usable with $(LREF enforce) which
+    expects exception constructors to take arguments in a fixed order. This
+    mixin provides that boilerplate code.
+
+    Note however that you need to mark the $(B mixin) line with at least a
+    minimal (i.e. just $(B ///) DDoc comment if you want the mixed-in
+    constructors to be documented in the newly created Exception subclass.
+ +/
+mixin template basicExceptionCtors()
+{
+    /++
+        Params:
+            msg  = The message for the exception.
+            file = The file where the exception occurred.
+            line = The line number where the exception occurred.
+            next = The previous exception in the chain of exceptions, if any.
+    +/
+    this(string msg, string file = __FILE__, size_t line = __LINE__,
+         Throwable next = null) @safe pure nothrow
+    {
+        super(msg, file, line, next);
+    }
+
+    /++
+        Params:
+            msg  = The message for the exception.
+            next = The previous exception in the chain of exceptions.
+            file = The file where the exception occurred.
+            line = The line number where the exception occurred.
+    +/
+    this(string msg, Throwable next, string file = __FILE__,
+         size_t line = __LINE__) @safe pure nothrow
+    {
+        super(msg, file, line, next);
+    }
+}
+
+///
+pure nothrow @safe unittest
+{
+    class MeaCulpa: Exception
+    {
+        ///
+        mixin basicExceptionCtors;
+    }
+
+    try
+        throw new MeaCulpa("test");
+    catch (Exception e) // Bugzilla 15467
+    {
+        assert(e.msg == "test");
```

Comment:
If you want the details of the PR, then you can look here: https://github.com/D-Programming-Language/druntime/pull/1445

But what it comes down to is that the function `message` has been added to `Throwable` with the idea that derived exception classes can do stuff like reuse a buffer to hold the message (it's `const(char)[]` rather than `string`), whereas `msg` is simply a `string`. If `message` is not overridden, then it just returns `msg`. So, in most cases, you'd get the same message regardless. The problem is that if you use` msg` on an object whose `message` function _has_ been overridden, then `msg` will be empty, and only `message` will have the actual message. So, any code which accesses `msg` on an exception should now call `message` in order to avoid problems when dealing with an exception class that overrides `message`. For now, `msg` has been left as-is, but we'll probably be deprecating it later so that folks don't end up with bugs where they used `msg` on an exception type which overrode `message` and thus end up with an empty string instead of the actual message.

So, your tests which access `msg` on a constructed exception should now call `message`, but none of the constructors need to change.